File: Components/DevExpressVCL/Demos/ExpressVerticalGrid/Delphi/StylesSimpleDemo/StylesSimpleDemoStylesDialog.pas
unit StylesSimpleDemoStylesDialog;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, cxStyles, cxTextEdit, cxInplaceContainer, cxControls,
  cxButtonEdit, cxClasses, StdCtrls, cxMRUEdit, cxMaskEdit,
  cxDBLookupComboBox, cxCurrencyEdit, cxMemo, cxCheckBox,
  cxLookAndFeelPainters, cxButtons, cxGraphics, cxCustomData, cxVGrid,
  cxEdit, cxEditRepositoryItems, ActnList;

type
  TStyles = (sBackground, sCategory, sContent, sHeader, sInactive, sIncSearch, sSelection);

  TStylesSimpleDemoStylesDialogForm = class(TForm)
    btnRestore: TcxButton;
    cxVerticalGrid: TcxVerticalGrid;
    cxEditRepository: TcxEditRepository;
    cxEditRepositoryMRUItem: TcxEditRepositoryMRUItem;
    ActionList1: TActionList;
    cxVerticalGridBackground: TcxEditorRow;
    cxVerticalGridCategory: TcxEditorRow;
    cxVerticalGridContent: TcxEditorRow;
    cxVerticalGridHeader: TcxEditorRow;
    cxVerticalGridInactive: TcxEditorRow;
    cxVerticalGridIncSearch: TcxEditorRow;
    cxVerticalGridSelection: TcxEditorRow;
    cxVerticalGridCaption: TcxMultiEditorRow;
    cxStyleRepository1: TcxStyleRepository;
    cxVerticalGridStyleSheetDevExpress: TcxVerticalGridStyleSheet;
    cxStyle1: TcxStyle;
    cxStyle2: TcxStyle;
    cxStyle3: TcxStyle;
    cxStyle4: TcxStyle;
    cxStyle5: TcxStyle;
    cxStyle6: TcxStyle;
    cxStyle7: TcxStyle;
    procedure FormCreate(Sender: TObject);
    procedure OnButtonClick(Sender: TObject);
    procedure btnRestoreClick(Sender: TObject);
    procedure OnEditValueChanged(Sender: TObject);
    procedure cxVerticalGridStylesGetContentStyle(Sender: TObject;
      AEditProp: TcxCustomEditorRowProperties; AFocused: Boolean;
      ARecordIndex: Integer; var AStyle: TcxStyle);
    procedure cxVerticalGridStylesGetHeaderStyle(Sender: TObject;
      ARow: TcxCustomRow; var AStyle: TcxStyle);
  private
    FRestoreDefaults: TNotifyEvent;
    function GetSelectedStyle: TcxStyle;
    procedure RefreshBinding;
    function GetCurrentStyle(AStyleID: TStyles): TcxStyle;
    procedure SetCurrentStyle(const AStyle: TcxStyle; AStyleID: TStyles);
  public
    property RestoreDefaults: TNotifyEvent read FRestoreDefaults write FRestoreDefaults;
  end;

var
  StylesSimpleDemoStylesDialogForm: TStylesSimpleDemoStylesDialogForm;

implementation

uses StylesSimpleDemoData, StylesSimpleDemoEdit, StylesSimpleDemoMain;

{$R *.dfm}

function ChangeStyleBinding(ACallback: TNotifyEvent): Boolean;
begin
  with TStylesSimpleDemoStylesDialogForm.Create(Application) do
  try
    RestoreDefaults := ACallback;
    Result := ShowModal = mrOK;
  finally
    Free;
  end;
end;

procedure TStylesSimpleDemoStylesDialogForm.FormCreate(Sender: TObject);
var
  i: integer;
  AStyle: TcxStyle;
begin
  OnClose := StylesSimpleDemoMainForm.StylesFormClosed;
  with cxEditRepositoryMRUItem.Properties.LookupItems, StylesSimpleDemoDataDM do
  begin
    Clear;
    for i := 0 to StyleRepository.Count - 1  do
    begin
       AStyle := TcxStyle(StyleRepository[i]);
       AddObject(AStyle.Name, AStyle);
    end;
  end;
  RefreshBinding;
end;

procedure TStylesSimpleDemoStylesDialogForm.OnButtonClick(
  Sender: TObject);
begin
  ChangeStyle(GetSelectedStyle);
end;

procedure TStylesSimpleDemoStylesDialogForm.btnRestoreClick(
  Sender: TObject);
begin
  if Assigned(FRestoreDefaults) then FRestoreDefaults(Sender);
  RefreshBinding;
end;

procedure TStylesSimpleDemoStylesDialogForm.RefreshBinding;
var
  I: Integer;
  AStyle: TcxStyle;
begin
  for i:=1 to cxVerticalGrid.Rows.Count - 1 do
  begin
    AStyle := GetCurrentStyle(TStyles(i-1));
    if AStyle <> nil then
      TcxEditorRow(cxVerticalGrid.Rows[i]).Properties.Value :=
        AStyle.Name;
  end;
end;

function TStylesSimpleDemoStylesDialogForm.GetSelectedStyle: TcxStyle;
var
  ItemIndex: Integer;
begin
  ItemIndex := cxEditRepositoryMRUItem.Properties.LookupItems.
    IndexOf(TcxEditorRow(cxVerticalGrid.FocusedRow).Properties.Value);
  Result := TcxStyle(cxEditRepositoryMRUItem.Properties.LookupItems.Objects[ItemIndex]);
end;

function TStylesSimpleDemoStylesDialogForm.GetCurrentStyle(AStyleID: TStyles): TcxStyle;
begin
  Result := nil;
  with StylesSimpleDemoMainForm do
  case AStyleID of
    sBackground:
      Result := cxDBVerticalGrid.Styles.Background;
    sCategory:
      Result := cxDBVerticalGrid.Styles.Category;
    sHeader:
      Result := cxDBVerticalGrid.Styles.Header;
    sContent:
      Result := cxDBVerticalGrid.Styles.Content;
    sInactive:
      Result := cxDBVerticalGrid.Styles.Inactive;
    sIncSearch:
      Result := cxDBVerticalGrid.Styles.IncSearch;
    sSelection:
       Result := cxDBVerticalGrid.Styles.Selection;
  end;
end;

procedure TStylesSimpleDemoStylesDialogForm.SetCurrentStyle(
  const AStyle: TcxStyle; AStyleID: TStyles);
begin
  with StylesSimpleDemoMainForm do
    case AStyleID of
      sBackground:
        cxDBVerticalGrid.Styles.Background := AStyle;
      sCategory:
        cxDBVerticalGrid.Styles.Category := AStyle;
      sHeader:
        cxDBVerticalGrid.Styles.Header := AStyle;
      sContent:
        cxDBVerticalGrid.Styles.Content := AStyle;
      sInactive:
        cxDBVerticalGrid.Styles.Inactive := AStyle;
      sIncSearch:
        cxDBVerticalGrid.Styles.IncSearch := AStyle;
      sSelection:
        cxDBVerticalGrid.Styles.Selection := AStyle;
    end;
end;

procedure TStylesSimpleDemoStylesDialogForm.OnEditValueChanged(
  Sender: TObject);
var
  ItemIndex: Integer;
  AStyle: TcxStyle;
begin
  ItemIndex := cxEditRepositoryMRUItem.Properties.LookupItems.IndexOf(TcxCustomMRUEdit(Sender).EditValue);
  AStyle := TcxStyle(cxEditRepositoryMRUItem.Properties.LookupItems.Objects[ItemIndex]);
  SetCurrentStyle(AStyle, TStyles(cxVerticalGrid.FocusedRow.VisibleIndex - 1));
end;

procedure TStylesSimpleDemoStylesDialogForm.cxVerticalGridStylesGetContentStyle(
  Sender: TObject; AEditProp: TcxCustomEditorRowProperties;
  AFocused: Boolean; ARecordIndex: Integer; var AStyle: TcxStyle);
begin
  if TcxEditorRow(AEditProp.Row).VisibleIndex = 0 then
    AStyle := cxVerticalGridStyleSheetDevExpress.Styles.Category
  else
    AStyle := cxVerticalGridStyleSheetDevExpress.Styles.Content;
end;

procedure TStylesSimpleDemoStylesDialogForm.cxVerticalGridStylesGetHeaderStyle(
  Sender: TObject; ARow: TcxCustomRow; var AStyle: TcxStyle);
begin
  if ARow.VisibleIndex = 0 then
    AStyle := cxVerticalGridStyleSheetDevExpress.Styles.Category
  else
    AStyle := cxVerticalGridStyleSheetDevExpress.Styles.Header;
end;

end.

File: Components/DevExpressVCL/Demos/ExpressVerticalGrid/CBuilder/StylesSimpleDemo/StylesSimpleDemoStylesDialog.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "StylesSimpleDemoStylesDialog.h"
#include "StylesSimpleDemoMain.h"
#include "StylesSimpleDemoData.h"
#include "StylesSimpleDemoEdit.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxButtons"
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxEdit"
#pragma link "cxEditRepositoryItems"
#pragma link "cxGraphics"
#pragma link "cxInplaceContainer"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxStyles"
#pragma link "cxVGrid"
#pragma resource "*.dfm"
TStylesSimpleDemoStylesDialogForm *StylesSimpleDemoStylesDialogForm;
//---------------------------------------------------------------------------
__fastcall TStylesSimpleDemoStylesDialogForm::TStylesSimpleDemoStylesDialogForm(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TStylesSimpleDemoStylesDialogForm::FormCreate(TObject *Sender)
{
  OnClose = StylesSimpleDemoMainForm->StylesFormClosed;
  TStrings *ALookupItems = cxEditRepositoryMRUItem->Properties->LookupItems;
  TcxStyleRepository *ARespository = StylesSimpleDemoDataDM->StyleRepository;
  ALookupItems->Clear();
  TcxStyle *AStyle;
  for (int I = 0; I < ARespository->Count; I++){
    AStyle = (TcxStyle*)ARespository->Items[I];
    ALookupItems->AddObject(AStyle->Name, AStyle);
  }
  RefreshBinding();
}
//---------------------------------------------------------------------------

void __fastcall TStylesSimpleDemoStylesDialogForm::OnButtonClick(TObject *Sender)
{
  ChangeStyle(GetSelectedStyle());
}
//---------------------------------------------------------------------------

void __fastcall TStylesSimpleDemoStylesDialogForm::btnRestoreClick(TObject *Sender)
{
  if (FRestoreDefaults != NULL) FRestoreDefaults(Sender);
  RefreshBinding();
}
//---------------------------------------------------------------------------

void __fastcall TStylesSimpleDemoStylesDialogForm::OnEditValueChanged(TObject *Sender)
{
  int ItemIndex = cxEditRepositoryMRUItem->Properties->LookupItems->IndexOf(
    ((TcxCustomMRUEdit*)Sender)->EditValue);
  TcxStyle *AStyle =
    (TcxStyle*)cxEditRepositoryMRUItem->Properties->LookupItems->Objects[ItemIndex];
  SetCurrentStyle(AStyle, (TStyles)(cxVerticalGrid->FocusedRow->VisibleIndex - 1));
}
//---------------------------------------------------------------------------

void __fastcall TStylesSimpleDemoStylesDialogForm::cxVerticalGridStylesGetContentStyle(
  TObject *Sender, TcxCustomEditorRowProperties *AEditProp, bool AFocused,
  int ARecordIndex, TcxStyle *&AStyle)
{
  if (((TcxEditorRow*)AEditProp->Row)->VisibleIndex == 0)
    AStyle = cxVerticalGridStyleSheetDevExpress->Styles->Category;
  else
    AStyle = cxVerticalGridStyleSheetDevExpress->Styles->Content;
}
//---------------------------------------------------------------------------

void __fastcall TStylesSimpleDemoStylesDialogForm::cxVerticalGridStylesGetHeaderStyle(
  TObject *Sender, TcxCustomRow *ARow, TcxStyle *&AStyle)
{
  if (ARow->VisibleIndex == 0)
    AStyle = cxVerticalGridStyleSheetDevExpress->Styles->Category;
  else
    AStyle = cxVerticalGridStyleSheetDevExpress->Styles->Header;
}
//---------------------------------------------------------------------------

TcxStyle* __fastcall TStylesSimpleDemoStylesDialogForm::GetSelectedStyle()
{
  int ItemIndex = cxEditRepositoryMRUItem->Properties->LookupItems->
    IndexOf(((TcxEditorRow*)cxVerticalGrid->FocusedRow)->Properties->Value);
  return (TcxStyle*)cxEditRepositoryMRUItem->Properties->LookupItems->Objects[ItemIndex];
}
//---------------------------------------------------------------------------

void __fastcall TStylesSimpleDemoStylesDialogForm::RefreshBinding()
{
  TcxStyle *AStyle;
  for (int I = 1; I < cxVerticalGrid->Rows->Count; I++){
    AStyle = GetCurrentStyle((TStyles)(I - 1));
    if (AStyle != NULL)
      ((TcxEditorRow*)cxVerticalGrid->Rows->Items[I])->Properties->Value =
        AStyle->Name;
  }
}
//---------------------------------------------------------------------------

TcxStyle* __fastcall TStylesSimpleDemoStylesDialogForm::GetCurrentStyle(
  TStyles AStyleID)
{
  TcxStyle *Result = NULL;
  TcxDBVerticalGrid *cxDBVerticalGrid = StylesSimpleDemoMainForm->cxDBVerticalGrid;
  switch (AStyleID){
    case sBackground:
      Result = cxDBVerticalGrid->Styles->Background;
      break;
    case sCategory:
      Result = cxDBVerticalGrid->Styles->Category;
      break;
    case sHeader:
      Result = cxDBVerticalGrid->Styles->Header;
      break;
    case sContent:
      Result = cxDBVerticalGrid->Styles->Content;
      break;
    case sInactive:
      Result = cxDBVerticalGrid->Styles->Inactive;
      break;
    case sIncSearch:
      Result = cxDBVerticalGrid->Styles->IncSearch;
      break;
    case sSelection:
      Result = cxDBVerticalGrid->Styles->Selection;
  }
  return Result;
}
//---------------------------------------------------------------------------

void __fastcall TStylesSimpleDemoStylesDialogForm::SetCurrentStyle(
  TcxStyle *AStyle, TStyles AStyleID)
{
  TcxDBVerticalGrid *cxDBVerticalGrid = StylesSimpleDemoMainForm->cxDBVerticalGrid;
  switch (AStyleID){
    case sBackground:
      cxDBVerticalGrid->Styles->Background = AStyle;
      break;
    case sCategory:
      cxDBVerticalGrid->Styles->Category = AStyle;
      break;
    case sHeader:
      cxDBVerticalGrid->Styles->Header = AStyle;
      break;
    case sContent:
      cxDBVerticalGrid->Styles->Content = AStyle;
      break;
    case sInactive:
      cxDBVerticalGrid->Styles->Inactive = AStyle;
      break;
    case sIncSearch:
      cxDBVerticalGrid->Styles->IncSearch = AStyle;
      break;
    case sSelection:
      cxDBVerticalGrid->Styles->Selection = AStyle;
      break;
  }
}
//---------------------------------------------------------------------------

